openssl/src/hash.rs

```diff
@@ -1,3 +1,36 @@
+//! This module provides an abstract interface over various specific cryptographic hash algorithms
+//! in the `Hasher` type. It is particularly useful in cases where you need to work with any type
+//! that supports a hashing interface. Furthermore, `Hasher` implements the [std::io::Write](
+//! https://doc.rust-lang.org/std/io/trait.Write.html) trait.
+//! # Examples
+//!
+//! ```rust
+//! extern crate openssl;
+//! extern crate hex;
+//!
+//! use std::io;
+//! use std::io::Write;
+//! use openssl::hash;
+//! use hex::ToHex;
+//!
+//! fn main() {
+//!     let mut hasher = hash::Hasher::new(hash::MessageDigest::sha512()).unwrap();
+//!     let bytes_hashed = hash_stream(&mut hasher).unwrap();
+//!     let hash = hasher.finish2().unwrap();
+//!     println!("Hashed {} bytes to {}", bytes_hashed, hash.to_hex());
+//! }
+//!
+//! fn hash_stream<W: Write>(hasher: &mut W) -> io::Result<usize> {
```

Comment:
I think it'd probably make more sense to have this take a `Hasher` and use `update` rather than `write`.

Reply:
I see your point, though I wanted to demonstrate the usefulness of `Write` being implemented.  With that in mind, would you still prefer just calling `update`?

Reply:
Having an example of why the Write implementation is useful is a good idea, but I don't think this does that - you can replace `W` with `Hasher` and `write` with `update` and nothing else really changes.

One option could be to use `io::copy` to hash the contents of a file.

Reply:
Duly noted! I think you'll like the changes I just pushed up.